src/main/java/com/google/sps/data/Room.java

```diff
@@ -18,16 +19,80 @@
     private static final String VIDEOS_PROPERTY = "videos";
+    public static final int MAX_MESSAGES = 10;
     private static DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     private List<Member> members;
-    private LinkedList<Message> messages;
+    private Queue<Message> messages;
     private Queue<Video> videos;
 
+    //Room factory function
+    public static Room createRoom(List<Member> members, Queue<Video> videos, Queue<Message> messages){
+        return new Room(members, videos, messages);
+
+    }
+
+    // Manipulates the messages property of the room given a room key and a new message to be added
+    public static void addMessagesFromKey(Key roomKey, Message chatMessage) {
+        try {
+        Entity roomEntity = datastore.get(roomKey);
+        Room room = Room.fromEntity(roomEntity);
+        room.addMessage(chatMessage);
+
+        roomEntity.setProperty("messages", room.getMessagesAsEntities());
+        datastore.put(roomEntity);
+        } catch (EntityNotFoundException e){
+            System.out.println(e.toString());
+        }
+    }
+
+    //Turns a Room entitiy into a Room object
+    public static Room fromEntity(Entity roomEntity) {
+        Map<String, Object> properties = roomEntity.getProperties();
+        List<Member> memberList = 
+        ((ArrayList<EmbeddedEntity>) properties.get(MEMBERS_PROPERTY)).stream().map(Member::fromEmbeddedEntity).collect(Collectors.toCollection(ArrayList::new));
+        Queue<Video> videoQueue = 
+        ((ArrayList<EmbeddedEntity>) properties.get(VIDEOS_PROPERTY)).stream().map(Video::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new));
+        Queue<Message> messageQueue = (Queue<Message>) properties.get(MESSAGES_PROPERTY) != null ?
+        ((ArrayList<EmbeddedEntity>) properties.get(MESSAGES_PROPERTY)).stream().map(Message::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new)) : new LinkedList();
+        return new Room(memberList, videoQueue, messageQueue);
+    }
+
+    //Turns the Room object into a datastore entity
+    public static Entity toEntity(Room room){
+        Entity newRoom = new Entity(ROOM_ENTITY);
+        newRoom.setProperty(MEMBERS_PROPERTY, room.getMembersAsEmbeddedEntities());
+        newRoom.setProperty(VIDEOS_PROPERTY, room.getVideosAsEntities());
+        newRoom.setProperty(MESSAGES_PROPERTY, room.getMessagesAsEntities());
+        return newRoom;
+    }
+
+    //Returns the Room's video url list
+    public Queue<Video> getVideos() {
+        return this.videos;
+    }
+
+    //Returns the Room's message list
+    public Queue<Message> getMessages(){
+        return this.messages;
+    }
+
+    //Returns the Room's members
+    public List<Member> getMembers() {
+        return this.members;
+    }
+
+    //Adds a video to the Room's video queue
+    public void addVideo(String url) {
+        this.videos.add(Video.createVideo(url));
+    }
```

Comment:
If you want to have adder files (like addVideo). Then the getter functions should create copies of the list fields. That way users of the room obj don't edit the content of videos/messages/members.